### PR TITLE
Update provision.sh to start the server with auth

### PR DESCRIPTION
**Pull request by Skarlso**
Label the PR with the kind of change this for:

kind/feature

**What this PR does / why we need it**:

Update provision.sh to start the server with auth.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes https://github.com/weaveworks-liquidmetal/flintlock/issues/539

**Special notes for your reviewer**:

**Checklist**:

- [ ] squashed commits into logical changes
- [ ] includes documentation
- [ ] adds unit tests
- [ ] adds or updates e2e tests

### CODE REVIEW

**Comment by Skarlso** [comment]
@Callisto13 I _think_ this is it, but is there any place else where these things have to be passed in?
Also, is it true that if you define tls_crt, you pretty much define the rest or should I check and add each individual setting one by one?

**Comment by Skarlso** [comment]
The Pr type doesn't like me. :(

File: hack/scripts/provision.sh [hunk 1]
```diff
@@ -754,6 +776,10 @@ cmd_all() {
 		"--dev")
 			DEVELOPMENT=true
 			;;
+		"--auth")
+		  shift
```
Comment (Skarlso) [inline hack/scripts/provision.sh:780]: Why is this is indented this weirdly?

Comment (Callisto13) [reply hack/scripts/provision.sh]: 🤷‍♀️ did you run a shell formatter over it? that should even everything out

File: hack/scripts/provision.sh [hunk 2]
```diff
@@ -788,7 +814,7 @@ cmd_all() {
 
 	install_firecracker "$fc_version"
 	do_all_containerd "$ctrd_version" "$set_thinpool"
-	do_all_flintlock "$fl_version" "$fl_address" "$fl_iface" "$bridge_name" "$insecure"
+	do_all_flintlock "$fl_version" "$fl_address" "$fl_iface" "$bridge_name" "$insecure", "$auth_file"
```
Comment (Callisto13) [inline hack/scripts/provision.sh:817]:
```suggestion
	do_all_flintlock "$fl_version" "$fl_address" "$fl_iface" "$bridge_name" "$insecure" "$auth_file"
```

File: hack/scripts/provision.sh (hunk 1 above)
Comment (Skarlso) [reply hack/scripts/provision.sh]: My IDE is crap. :( It didn't insert the right indentation character. I do have shell checker, but not formatter, I think. I shall install it. :)

File: hack/scripts/provision.sh [hunk 3]
```diff
@@ -341,6 +342,25 @@ write_flintlockd_config() {
...
+    if [[ -n "$tls_cert" ]]; then
+      # shellcheck disable=SC2154
+      auth_information=<< TLS
+tls-cert="$tls_cert"
+tls-key="$tls_key"
+tls-client-validate="$tls_client_validate"
+tls-client-ca="$tls_client_ca"
+TLS
+    fi
+	fi
```
Comment (Callisto13) [inline hack/scripts/provision.sh:363]: can you run a formatter over this? the spacing is very weird

File: hack/scripts/provision.sh (hunk 2 above)
Comment (Skarlso) [reply hack/scripts/provision.sh:817]: Ah good ol' bash. Got too used to Go. :D

File: hack/scripts/provision.sh (hunk 3 above)
Comment (Skarlso) [reply hack/scripts/provision.sh:363]: :nod: will do.

**Comment by Callisto13** [comment]
can you update the help menus? and screenshot/paste in some successful calls?

File: hack/scripts/provision.sh [hunk 4]
```diff
@@ -355,15 +358,36 @@ insecure: $insecure
 EOF
 
 	if [[ -n "$bridge_name" ]]; then
-	cat <<EOF >>"$FLINTLOCKD_CONFIG_PATH"
+		cat <<EOF >>"$FLINTLOCKD_CONFIG_PATH"
```
Comment (Skarlso) [inline hack/scripts/provision.sh:361]: @Callisto13 I ran shall-formatter 🤷

**Comment by Skarlso** [comment]
I updated the help and the pass and also added it to the flintlock command, not just all. :)

I will work on adding screenshots tomorrow.

**Comment by Skarlso** [comment]
That is also possible indeed. And would reduce the burden on the script. That's a good thought!

**Comment by Skarlso** [comment]
The only thing that might be a problem is if they define something that we already defined? Such as `verbosity` for example. Or should we be strict that you should only define value x, y and z? Or should we try to do a merge... we could use [paste](https://ss64.com/bash/paste.html) for that.

**Comment by Skarlso** [comment]
@Callisto13 I made an attempt to just use the thing and paste it after the things we would set up. For deduplication, see my above comment. :)

**Comment by Callisto13** [comment]
> The only thing that might be a problem is if they define something that we already defined?

In an ideal situation they should be able to define what they like... but implementing that is tricky... at least in bash 😄

**Comment by Skarlso** [comment]
```
root@ubuntu2004:~# ./provision.sh flintlock --dev --insecure --bridge lmbr0 --grpc-address 0.0.0.0:9090 -f flintlock.config
[flintlock provision.sh] Creating containerd directory /var/lib/containerd-dev/snapshotter/devmapper
[flintlock provision.sh] Creating containerd directory /run/containerd-dev
[flintlock provision.sh] Creating containerd directory /etc/containerd
[flintlock provision.sh] All containerd directories created
[flintlock provision.sh] Installing flintlockd version latest to /usr/local/bin
[flintlock provision.sh] Flintlockd version v0.4.0 successfully installed
[flintlock provision.sh] Writing flintlockd config to /etc/opt/flintlockd/config.yaml.
[flintlock provision.sh] merging provided config file with the created one
[flintlock provision.sh] Flintlockd config saved
[flintlock provision.sh] Starting flintlockd service with /etc/systemd/system/flintlockd.service
[flintlock provision.sh] Flintlockd running at 0.0.0.0:9090 via interface eth0
```

Config file for flintlockd:

```
cat /etc/opt/flintlockd/config.yaml
---
insecure: false
grpc-endpoint: 0.0.0.0:9090
tls-cert: /root/tls-cert.crt
verbosity: 9
bridge-name: lmbr0
tls-key: /root/tls-key.key
containerd-socket: /run/containerd-dev/containerd.sock

root@ubuntu2004:~#
```

**Comment by Skarlso** [comment]
Interesting bit that hammertime did not return an auth failed, but rather something weird:

```
systemctl stop flintlockd.service
root@ubuntu2004:~# ./hammertime list -a 0.0.0.0:9090
2022/10/19 10:52:23 rpc error: code = Unavailable desc = connection error: desc = "transport: Error while dialing dial tcp 0.0.0.0:9090: connect: connection refused"
root@ubuntu2004:~# systemctl start flintlockd.service
root@ubuntu2004:~# ./hammertime list -a 0.0.0.0:9090
2022/10/19 10:53:23 rpc error: code = Unavailable desc = connection closed before server preface received
```

Log from flintlockd:

```
root@ubuntu2004:~# systemctl status flintlockd.service
● flintlockd.service - flintlock microvm service
     Loaded: loaded (/etc/systemd/system/flintlockd.service; enabled; vendor preset: enabled)
     Active: active (running) since Wed 2022-10-19 10:43:13 UTC; 7min ago
       Docs: https://docs.flintlock.dev/
    Process: 28709 ExecStartPre=/usr/bin/which firecracker (code=exited, status=0/SUCCESS)
    Process: 28710 ExecStartPre=/usr/bin/which flintlockd (code=exited, status=0/SUCCESS)
   Main PID: 28711 (flintlockd)
      Tasks: 7 (limit: 9443)
     Memory: 10.2M
     CGroup: /system.slice/flintlockd.service
             └─28711 /usr/local/bin/flintlockd run

Oct 19 10:43:13 ubuntu2004.localdomain flintlockd[28711]: time="2022-10-19T10:43:13Z" level=info msg="flintlockd grpc api server starting"
Oct 19 10:43:13 ubuntu2004.localdomain flintlockd[28711]: time="2022-10-19T10:43:13Z" level=info msg="starting microvm controller"
Oct 19 10:43:13 ubuntu2004.localdomain flintlockd[28711]: time="2022-10-19T10:43:13Z" level=info msg="starting microvm controller with 1 workers" controller=microvm
Oct 19 10:43:13 ubuntu2004.localdomain flintlockd[28711]: time="2022-10-19T10:43:13Z" level=info msg="resyncing microvm specs" controller=microvm
Oct 19 10:43:13 ubuntu2004.localdomain flintlockd[28711]: time="2022-10-19T10:43:13Z" level=trace msg="querying all microvms: map[Namespace:]" component=app controller=microvm
Oct 19 10:43:13 ubuntu2004.localdomain flintlockd[28711]: time="2022-10-19T10:43:13Z" level=warning msg="basic authentication is DISABLED"
Oct 19 10:43:13 ubuntu2004.localdomain flintlockd[28711]: time="2022-10-19T10:43:13Z" level=info msg="TLS is enabled"
Oct 19 10:43:13 ubuntu2004.localdomain flintlockd[28711]: time="2022-10-19T10:43:13Z" level=debug msg="starting grpc server listening on endpoint 0.0.0.0:9090"
Oct 19 10:43:13 ubuntu2004.localdomain flintlockd[28711]: time="2022-10-19T10:43:13Z" level=info msg="starting event listener" controller=microvm
Oct 19 10:43:13 ubuntu2004.localdomain flintlockd[28711]: time="2022-10-19T10:43:13Z" level=info msg="Starting workersnum_workers1" controller=microvm

```

It _is_ running. 🤷

**Comment by Skarlso** [comment]
Re-installed without TLS to show that without specifying a config it works normally:

```
./provision.sh flintlock --dev --insecure --bridge lmbr0 --grpc-address 0.0.0.0:9090
[flintlock provision.sh] Creating containerd directory /var/lib/containerd-dev/snapshotter/devmapper
[flintlock provision.sh] Creating containerd directory /run/containerd-dev
[flintlock provision.sh] Creating containerd directory /etc/containerd
[flintlock provision.sh] All containerd directories created
[flintlock provision.sh] Installing flintlockd version latest to /usr/local/bin
[flintlock provision.sh] Flintlockd version v0.4.0 successfully installed
[flintlock provision.sh] Writing flintlockd config to /etc/opt/flintlockd/config.yaml.
[flintlock provision.sh] Flintlockd config saved
[flintlock provision.sh] Starting flintlockd service with /etc/systemd/system/flintlockd.service
[flintlock provision.sh] Flintlockd running at 0.0.0.0:9090 via interface eth0
root@ubuntu2004:~# ./hammertime list -a 0.0.0.0:9090
{}
root@ubuntu2004:~#
```

**Comment by Skarlso** [comment]
Okay, I explored everything I could with this. I even enabled validation, generate the CA cert, did everything and got that flintlock is up and running:

```
systemctl status flintlockd.service
● flintlockd.service - flintlock microvm service
     Loaded: loaded (/etc/systemd/system/flintlockd.service; enabled; vendor preset: enabled)
     Active: active (running) since Thu 2022-10-20 08:09:16 UTC; 3s ago
       Docs: https://docs.flintlock.dev/
    Process: 4503 ExecStartPre=/usr/bin/which firecracker (code=exited, status=0/SUCCESS)
    Process: 4504 ExecStartPre=/usr/bin/which flintlockd (code=exited, status=0/SUCCESS)
   Main PID: 4505 (flintlockd)
      Tasks: 7 (limit: 9443)
     Memory: 10.4M
     CGroup: /system.slice/flintlockd.service
             └─4505 /usr/local/bin/flintlockd run

Oct 20 08:09:16 ubuntu2004.localdomain flintlockd[4505]: time="2022-10-20T08:09:16Z" level=info msg="flintlockd grpc api server starting"
Oct 20 08:09:16 ubuntu2004.localdomain flintlockd[4505]: time="2022-10-20T08:09:16Z" level=info msg="starting microvm controller"
Oct 20 08:09:16 ubuntu2004.localdomain flintlockd[4505]: time="2022-10-20T08:09:16Z" level=info msg="starting microvm controller with 1 workers" controller=microvm
Oct 20 08:09:16 ubuntu2004.localdomain flintlockd[4505]: time="2022-10-20T08:09:16Z" level=info msg="resyncing microvm specs" controller=microvm
Oct 20 08:09:16 ubuntu2004.localdomain flintlockd[4505]: time="2022-10-20T08:09:16Z" level=trace msg="querying all microvms: map[Namespace:]" component=app controller=microvm
Oct 20 08:09:16 ubuntu2004.localdomain flintlockd[4505]: time="2022-10-20T08:09:16Z" level=warning msg="basic authentication is DISABLED"
Oct 20 08:09:16 ubuntu2004.localdomain flintlockd[4505]: time="2022-10-20T08:09:16Z" level=info msg="TLS is enabled"
Oct 20 08:09:16 ubuntu2004.localdomain flintlockd[4505]: time="2022-10-20T08:09:16Z" level=info msg="starting event listener" controller=microvm
Oct 20 08:09:16 ubuntu2004.localdomain flintlockd[4505]: time="2022-10-20T08:09:16Z" level=info msg="Starting workersnum_workers1" controller=microvm
Oct 20 08:09:16 ubuntu2004.localdomain flintlockd[4505]: time="2022-10-20T08:09:16Z" level=debug msg="starting grpc server listening on endpoint 0.0.0.0:9090"
```

With config:

```
root@ubuntu2004:~# cat flintlock.config
tls-cert: /root/service.pem
tls-key: /root/service.key
tls-client-validate: true
tls-client-ca: /root/ca-cert.pem
```

I checked hammertime it is just forwarding the response. The interesting part is that flintlock doesn't log ANYTHING. It's as if hammertime can't even see it.

Maybe there is something wrong with the port it will run on?

config looks ok to me:

```
---
insecure: false
grpc-endpoint: 0.0.0.0:9090
tls-cert: /root/service.pem
tls-client-ca: /root/ca-cert.pem
tls-client-validate: true
verbosity: 9
bridge-name: lmbr0
tls-key: /root/service.key
containerd-socket: /run/containerd-dev/containerd.sock
```

**Comment by Callisto13** [comment]
I am happy with what you have pasted here. The fault lies with hammertime probably. 

> The interesting part is that flintlock doesn't log ANYTHING. It's as if hammertime can't even see it.

I _think_ i recall flintlock not logging failed auth attempts before when i was implementing this. I believe it happens at a different level (ie at the protocol level, not somewhere in flintlock itself) and we are not exposing any logs there (idek if we could, but might be nice to know who is _trying_ to connect? something for the future)

hopefully i will get to review this soon ty :tada:

**Comment by Skarlso** [comment]
Oh yeah, that makes sense. Cool! :) Thanks. :)

File: hack/scripts/provision.sh [hunk 5]
```diff
@@ -995,15 +1024,16 @@ cmd_all_help() {
...
+      -y                          Autoapprove all prompts (danger)
+      --skip-apt, -s              Skip installation of apt packages
+      --thinpool, -t              Name of thinpool to create (default: flintlock or flintlock-dev)
+      --disk, -d                  Name blank unpartioned disk to use for direct lvm thinpool (ignored if --dev set)
+      --grpc-address, -a          Address on which to start the Flintlock GRPC server (default: local ipv4 address)
+      --parent-iface, -i          Interface of the default route of the host
+      --bridge, -b                Bridge to use instead of an interface (will override --parent-iface)
+      --insecure, -k              Start flintlockd without basic auth or certs
+      --dev                       Set up development environment. Loop thinpools will be created.
+      --flintlock-config-file, -f Provide a configuration file to set up authentication for flintlock such as, token or TLS
```
Comment (Callisto13) [inline hack/scripts/provision.sh:1036]:
```suggestion
      --flintlock-config-file, -f Path to a valid flintlockd configuration file with overriding config
```

File: hack/scripts/provision.sh [hunk 6]
```diff
@@ -1038,6 +1068,7 @@ cmd_flintlock_help() {
       --bridge, -b       Bridge to use instead of an interface (will override --parent-iface)
       --insecure, -k     Start flintlockd without basic auth or certs
       --dev              Assumes containerd has been provisioned in a dev environment
+      --config-file, -f  Provide a configuration file to set up authentication for flintlock such as, token or TLS (the file should follow shell syntax)
```
Comment (Callisto13) [inline hack/scripts/provision.sh:1071]:
```suggestion
      --config-file, -f  Path to a valid flintlockd configuration file with overriding config
```

**Comment by Callisto13** [comment]
what happens if they provide a config file with the yaml `---` at the top of it? does it get processed okay?

**Comment by Skarlso** [comment]
Good question. I didn't realise that the settings were yaml hahaha. I think it would blow up but I'll test it.

**Comment by Callisto13** [comment]
We could probably just have a sed to remove anything like that before processing

**Comment by Skarlso** [comment]
@Callisto13 Heh, it produces this:

```yaml
---
insecure: false
grpc-endpoint: 0.0.0.0:9090
tls-cert: /root/service.pem
tls-client-ca: /root/ca-cert.pem
---:
tls-client-validate: true
verbosity: 9
bridge-name: lmbr0
tls-key: /root/service.key
containerd-socket: /run/containerd-dev/containerd.sock
```

Incidentally, flintlock actually didn't care and started up anyways hahaha.

That said, I'll put in a check to skip anything that doesn't contain a `:` in it.

**Comment by Skarlso** [comment]
The new fix generated the right config file:
```yaml
---
insecure: false
grpc-endpoint: 0.0.0.0:9090
tls-cert: /root/service.pem
tls-client-ca: /root/ca-cert.pem
tls-client-validate: true
verbosity: 9
bridge-name: lmbr0
tls-key: /root/service.key
containerd-socket: /run/containerd-dev/containerd.sock
```

flintlock config file:
```yaml
cat flintlock.config
---
tls-cert: /root/service.pem
tls-key: /root/service.key
tls-client-validate: true
tls-client-ca: /root/ca-cert.pem
```

**Comment by Skarlso** [comment]
@Callisto13 Shall I squash the commits or will you with Github? :)

**Comment by Callisto13** [comment]
i can when i merge 👍

File: hack/scripts/provision.sh [hunk 7]
```diff
@@ -343,29 +345,49 @@ write_flintlockd_config() {
...
+		settings["bridge-name"]="$bridge_name"
 	else
-	cat <<EOF >>"$FLINTLOCKD_CONFIG_PATH"
-parent-iface: "$parent_iface"
-EOF
+		settings["parent-iface"]="$parent_iface"
+	fi
+
+	if [[ -n "$config_file" ]]; then
+		say "merging provided config file with the created one"
```
Comment (Callisto13) [inline hack/scripts/provision.sh:367]:
```suggestion
		say "Merging provided flintlockd config file with auto-generated options"
```

**Comment by Skarlso** [comment]
> i can when i merge 👍

Thanks! :)

Also, this was quite a cool journey! :)

**Review by Callisto13 — APPROVED** [review]

**Pull request merged**